libraries/GCS_MAVLink/GCS_FTP.cpp

```diff
@@ -0,0 +1,600 @@
+    ftp.replies = nullptr;
+
+    return false;
+    
+}
+
+void GCS_MAVLINK::handle_file_transfer_protocol(const mavlink_message_t &msg) {
+    if(ftp_init()) {
+        mavlink_file_transfer_protocol_t packet;
+        mavlink_msg_file_transfer_protocol_decode(&msg, &packet);
+
+        struct pending_ftp request;
+
+        request.chan = chan;
+        request.seq_number = *(uint16_t *)packet.payload;
+
+        request.session = packet.payload[2];
+        request.opcode = static_cast<FTP_OP>(packet.payload[3]);
+        request.size = packet.payload[4];
+        request.req_opcode = static_cast<FTP_OP>(packet.payload[5]);
+        request.burst_complete = packet.payload[6];
+        request.offset = *(uint32_t *)(&packet.payload[8]);
+        request.sysid = msg.sysid;
+        request.compid = msg.compid;
+        memcpy(request.data, &packet.payload[12], sizeof(packet.payload) - 12);
+
+        if (!ftp.requests->push(request)) {
+            // dropping the message, no buffer space to queue it in
+            // we could NACK it, but that can lead to GCS confusion, so we're treating it like lost data
+        }
+    }
+}
+
+void GCS_MAVLINK::send_ftp_replies(void) {
+    if (ftp.replies != nullptr) {
+        for (uint8_t i = 0; i < 20; i++) {
+            struct pending_ftp reply;
+            uint8_t payload[251] = {};
+            if (ftp.replies->peek(reply) && (reply.chan == chan)) {
+                    if ((i > 0) && comm_get_txspace(chan) < (2 * (packet_overhead() + MAVLINK_MSG_ID_FILE_TRANSFER_PROTOCOL_LEN))) {
+                        // if this isn't the first packet we have to leave deadspace for the next message
+                        break;
+                    }
+                    if (!HAVE_PAYLOAD_SPACE(chan, FILE_TRANSFER_PROTOCOL)) {
+                        return;
+                    }
+                    ((uint16_t *)payload)[0] = reply.seq_number;
+                    payload[2] = reply.session;
+                    payload[3] = static_cast<uint8_t>(reply.opcode);
+                    payload[4] = reply.size;
+                    payload[5] = static_cast<uint8_t>(reply.req_opcode);
+                    payload[6] = reply.burst_complete ? 1 : 0;
+                    *(uint32_t *)(&payload[8]) = reply.offset;
+                    memcpy(&payload[12], reply.data, sizeof(reply.data));
+                    mavlink_msg_file_transfer_protocol_send(
+                        reply.chan,
+                        0, reply.sysid, reply.compid,
+                        payload);
+                    ftp.replies->pop(reply);
+            } else {
+                break;
+            }
+        }
+    }
+}
+
+void GCS_MAVLINK::ftp_error(struct pending_ftp &response, FTP_ERROR error) {
+    response.opcode = FTP_OP::Nack;
+    response.data[0] = static_cast<uint8_t>(error);
+    response.size = 1;
+
+    // FIXME: errno's are not thread-local as they should be on ChibiOS
+    if (error == FTP_ERROR::FailErrno) {
```

Comment:
Our pattern for enum classes would say `FTPError` rather than `FTP_ERROR`.  If there's an external document, please reference it next to the `enum class` declarations

Reply:
Can we compromise on `FTP_Error` instead? I don't like `FTPError` as it runs into the acronym and gets harder to read.